### ISSUE TITLE
proposal: add simplify (-s) flag to shfmt

### ISSUE THREAD

**Issue by mvdan** [mvdan/sh#34]
Much like `gofmt -s`, which removes bits of code that aren't necessary such as repeated types.

Rules that come to mind, which would be useful in the shell:

* Get rid of backticks - `` `foo` `` -> `$(foo)`
* Remove useless word parts - `foo""bar -> foobar` (applies to `""`, `''`, `$()` and perhaps more)
* Force functions to use blocks - `foo() bar` -> `foo() { bar; }` (the first works under some shells)

Some more rules that might be too pedantic/opinionated:

* Force portable functions - `function foo() { bar; }` -> `foo() { bar; }`
* Replace `${foo}` with `$foo` where it means the same
* Remove quotes where they're unnecessary - `"foo"` -> `foo`

Feedback and ideas welcome @focusaurus @ebfe @fwip @mmlb @legionus @mkovacs @seletskiy

**Comment by mvdan** [comment]
Note that being a linter or rewriting/restructuring code is outside the scope of this. Only very simple rules would apply.

**Comment by focusaurus** [comment]
All of those sound useful to me. I'd like each to have a clear name, all the names to follow a consistent convention, and be able to easily toggle them on/off for a whole run, and maybe for specific occurrences. I think eslint in the javascript world does a pretty good job of this, at least from a linting perspective.

**Comment by focusaurus** [comment]
I'd predict a lot of discussion if this project becomes widely used around "very simple" so I'd recommend thinking about and clearly documenting your criteria for what is in scope as a code mod and what is out of scope. For example, if I were to ask for variable naming convention enforcement, is that "simple"?

**Comment by focusaurus** [comment]
Loosely related, [shellcheck](https://github.com/koalaman/shellcheck) is a linter, but all it's rules have opaque numeric codes which are a lot harder to work with than word-based names.

**Comment by mvdan** [comment]
> be able to easily toggle them on/off

Ideally they would only be all on or off at once, i.e. whether or not `-s` was given. This is why I want to think about what rules would be too intrusive, as I'd rather not have switches for each one of them separately.

> clearly documenting your criteria for what is in scope as a code mod and what is out of scope

This is a formatter, so only simplifications in format. In other words, anything that can be rewritten in a simpler or better way, as long as the logic is simple.

> For example, if I were to ask for variable naming convention enforcement, is that "simple"?

That's style, not format. In other words, this would have to output warnings instead of rewriting the file as a computer can't apply a naming convention easily. It changes the program, not just printing it in a simpler format.

**Comment by mvdan** [comment]
I've decided not to go forward with this, neither with `-s` nor with `ast.Simplify()`. I realised the only change that would make sense is the switch of `` `foo` `` for `$(foo)`. The rest are bad for various reasons:

> Remove useless word parts - `foo""bar -> foobar` (applies to `""`, `''`, `$()` and perhaps more)

This is very rare. It's also very hard to do properly. For example, `$foo""bar` has three word parts, but `$foobar` has one and means a different thing. This would be very tricky and it would be hard to check that we're never breaking programs.

> Force functions to use blocks - `foo() bar` -> `foo() { bar; }` (the first works under some shells)

This is very weird and not even supported by most shells, so it's not a feature for anyone.

> Force portable functions - `function foo() { bar; }` -> `foo() { bar; }`
> Replace `${foo}` with `$foo` where it means the same
> Remove quotes where they're unnecessary - `"foo"` -> `foo`

I do think the first one would be useful, the other ones not so much. After seeing some projects uniformly use `function`, unnecessary `${}` and unnecessary quotes, I think people would mostly just be bothered by these changes. Sure they remove bytes, but sometimes they might not help with readability, consistency or other more important aspects.

As such, I'll simply make `shfmt` always replace `` `foo` `` with `$(foo)`. This will be done at the parser level - both will parse as the same, and the printer will always use the `$(foo)` form.
